File: functions/api/recap.js
// GET /api/recap?date=YYYY-MM-DD&r=<token> -- somebody else's finished game:
// their five pins, the five answers, and how far apart they were.
//
// The share string is deliberately spoiler-free -- five coloured squares that
// say how a game went and nothing about where it was. This is the other half,
// for after: a player who wants to show a friend *how* they got 23,440 has
// nothing to send but a screenshot, and a screenshot cannot be zoomed into or
// argued with.
//
// Two rules keep it from being a way to read ahead:
//
//   The date must be closed. Everything here is an answer key, so a link to a
//   day still in play would spoil it for whoever opens it -- the one person the
//   sender was not thinking about. Closed everywhere, not merely over where the
//   sender lives; the play window spans UTC+14 to UTC-12.
//
//   The token names one player on one date, and is a hash rather than their id
//   (see _recap.mjs). Nothing here can be turned back into the credential that
//   would let the reader play as them.
//
// What it will not do is invent history: a game recorded before guesses were
// stored comes back with its scores and null coordinates, and the page draws
// those rounds without a line. Showing a plausible pin would be worse than
// showing none.
import { isClosed, playWindow } from '../../web/daily.js';
import { json, DATE } from '../_json.mjs';
import { nameExpr, PLACEHOLDER } from '../_names.mjs';
import { playerFor } from '../_recap.mjs';

// A finished day cannot change -- its schedule is frozen, its plays are
// first-write-wins, and it is past its close -- so a recap of one is the same
// bytes forever. The only thing that can move is the player's name.
const CACHE = { 'cache-control': 'public, max-age=3600' };

export async function onRequestGet({ request, env }) {
  const params = new URL(request.url).searchParams;
  const date = params.get('date'), token = params.get('r') || '';
  if (!date || !DATE.test(date)) {
    return json({ error: 'expected ?date=YYYY-MM-DD&r=<token>' }, 400);
  }

  // `closes` goes back with the refusal so the page can say how long the wait
  // is rather than only that there is one. A future date has a close too, and it
  // is honest for the same reason -- that game has not been played yet either.
  if (!isClosed(date)) {
    return json({
      error: 'that day is still being played',
      closes: playWindow(date).closes,
    }, 403);
  }

  const playerId = await playerFor(env, date, token);
  // One message for a token that matches nobody, whether it was mistyped,
  // truncated by whatever pasted it, or belongs to a player who has since been
  // merged into another id by /api/link. The reader can do nothing about any of
  // them except ask for the link again.
  if (!playerId) return json({ error: 'no game found for that link' }, 404);

  // round_days first, so the rounds come back in the order they were played and
  // the numbers on the map match the ones on the contact sheet. An inner join to
  // plays drops the rounds this player never answered -- a recap shows the game
  // they had, which is sometimes three rounds and a night's sleep.
  const { results: rounds } = await env.ANSWERS
    .prepare(`SELECT d.image, a.lat, a.lng, a.state, a.filmed,
                     p.km, p.points, p.guess_lat, p.guess_lng
                FROM round_days d
                JOIN plays p ON p.date = d.date AND p.image = d.image AND p.player_id = ?
                JOIN answers a ON a.image = d.image
               WHERE d.date = ?
               ORDER BY d.position`)
    .bind(playerId, date)
    .all();

  // playerFor only answers with a player that has rows on this date, so an empty
  // result means the schedule went missing underneath them rather than that the
  // link was wrong -- a round deleted from round_days, or answers never seeded.
  if (!rounds.length) return json({ error: 'that game cannot be rebuilt' }, 404);

  // A bare SELECT with no FROM: the whole statement is the name expression, so
  // the shared rule needs no table to hang off and ?1 is bound once for both
  // halves of it.
  const named = await env.ANSWERS
    .prepare(`SELECT ${nameExpr('?1')} AS name`)
    .bind(playerId)
    .first();

  return json({
    date,
    name: named?.name || PLACEHOLDER,
    total: rounds.reduce((sum, r) => sum + r.points, 0),
    rounds,
  }, 200, CACHE);
}

File: server/recap.py
"""POST /api/games and GET /api/recap -- a player's own finished dailies, and a
shareable view of one of them.

functions/api/games.js, functions/api/recap.js and functions/_recap.mjs carry
the reasoning; this is the same contract. In short: the player id is a
credential, so a share link carries a date-salted, truncated hash of it instead,
and nothing reveals a day's answers until that day is closed everywhere.
"""

import hashlib
import re

from server import rules
from server.leaderboard import PLACEHOLDER, name_expr

# A month of daily games: long enough to find the one worth showing somebody,
# short enough that the response stays a list rather than an archive.
GAMES = 30
TOKEN_BYTES = 6
TOKEN = re.compile(rf"[0-9a-f]{{{TOKEN_BYTES * 2}}}")
# A finished day cannot change, so a recap of one is the same bytes forever bar
# the player's name.
CACHE = {"cache-control": "public, max-age=3600"}


def token_for(player_id: str, date: str) -> str:
    digest = hashlib.sha256(f"{date}:{player_id}".encode()).digest()
    return digest[:TOKEN_BYTES].hex()


def is_closed(date: str, now=None) -> bool:
    """Done being played anywhere on Earth. Not the negation of is_open: a date
    next week is neither, and a recap of it would spoil a game nobody has had."""
    return date <= rules.last_closed_date(now)


async def player_for(db, date: str, token: str) -> str | None:
    # ponytail: linear in that date's players, which is tens; store the token as
    # a column (backfilled from token_for) if a day ever draws thousands.
    if not TOKEN.fullmatch(token):
        return None
    rows = await db.fetchall(
        "SELECT DISTINCT player_id FROM plays WHERE date = ?", date
    )
    for row in rows:
        if token_for(row["player_id"], date) == token:
            return row["player_id"]
    return None


async def games(db, body, now=None) -> tuple:
    player_id = body.get("player_id") if isinstance(body, dict) else None
    if not rules.is_player_id(player_id):
        return 400, {"error": "expected {player_id}"}
    rows = await db.fetchall(
        """SELECT date, SUM(points) AS total, COUNT(*) AS rounds
             FROM plays
            WHERE player_id = ?
            GROUP BY date
            ORDER BY date DESC
            LIMIT ?""",
        player_id,
        GAMES,
    )
    # A token only for a day closed everywhere: a link to one still open is a
    # spoiler for whoever it is sent to.
    listed = [
        {
            **g,
            "token": token_for(player_id, g["date"])
            if is_closed(g["date"], now)
            else None,
        }
        for g in rows
    ]
    return 200, {"games": listed}, {"cache-control": "no-store"}


async def recap(db, params, now=None) -> tuple:
    date, token = params.get("date"), params.get("r") or ""
    if not rules.is_calendar_date(date):
        return 400, {"error": "expected ?date=YYYY-MM-DD&r=<token>"}
    if not is_closed(date, now):
        # Milliseconds, as the page's playWindow() counts them.
        closes = rules.play_window(date)[1]
        return 403, {
            "error": "that day is still being played",
            "closes": int(closes.timestamp() * 1000),
        }

    player_id = await player_for(db, date, token)
    if not player_id:
        return 404, {"error": "no game found for that link"}

    # In the order they were played; the inner join to plays drops rounds this
    # player never answered.
    rounds = await db.fetchall(
        """SELECT d.image, a.lat, a.lng, a.state, a.filmed,
                  p.km, p.points, p.guess_lat, p.guess_lng
             FROM round_days d
             JOIN plays p ON p.date = d.date AND p.image = d.image AND p.player_id = ?
             JOIN answers a ON a.image = d.image
            WHERE d.date = ?
            ORDER BY d.position""",
        player_id,
        date,
    )
    if not rounds:
        return 404, {"error": "that game cannot be rebuilt"}

    named = await db.fetchone(f"SELECT {name_expr('?1')} AS name", player_id)
    return (
        200,
        {
            "date": date,
            "name": (named or {}).get("name") or PLACEHOLDER,
            "total": sum(r["points"] for r in rounds),
            "rounds": rounds,
        },
        CACHE,
    )
